feat(bb.js): remove Pino dependency and implement platform-agnostic logger

johnathan79717 opened the pull request:
## Summary
Removes the Pino logging dependency from bb.js and replaces it with a simple console-based logger that outputs to stderr.

## Changes
- Simplified logger to use `console.error` (stderr) with name prefixes
- Removed complex log level management and worker thread log options
- Removed Pino dependency from package.json
- Users can still provide custom logger via `BackendOptions.logger`
- Updated browser test app to output bb.js logs to both pino and console for CI benchmark parsing

## Implementation
```typescript
export function createDebugLogger(name: string): (msg: string) => void {
  return (msg: string) => {
    console.error(`[${name}] ${msg}`);
  };
}
```

## Why stderr?
Logger output must go to stderr to avoid interfering with scripts that parse stdout:
```bash
aztec get-canonical-sponsored-fpc-address | awk '{print $NF}'
```

If logs went to stdout, they would be captured by `awk` instead of the actual command output. Using stderr:
- Prevents interference with stdout-based command pipelines
- Scripts parsing stdout work correctly
- Logs still visible in terminal
- Can be suppressed with `2>/dev/null`

## Browser Test App Logger
The browser test app demonstrates how users can provide their own logger implementation. The `bbLogger` wrapper:
- Logs to pino for structured logging in the application
- Also outputs to console.log for CI benchmark scripts that parse memory usage from plain-text logs
- This dual approach supports both application needs and CI tooling requirements

## Testing
- ✅ All tests pass (26 tests)
- ✅ Browser integration test passes
- ✅ Builds succeed for ESM, CJS, and browser targets

Contributes to https://github.com/AztecProtocol/barretenberg/issues/1590

johnathan79717 committed:
feat(bb.js): simplify logger to use console.log by default

Simplify the custom logger implementation to use console.log by default
instead of complex log level management. Users can still pass their own
logger function via BackendOptions.

Changes:
- Replace complex logger with simple console.log-based implementation
- Remove LogLevels and LogOptions types
- Remove log option passing to worker threads
- Maintain ability to provide custom logger via BackendOptions.logger

Contributes to #1590

ludamad commented on barretenberg/ts/src/log/logger.ts:
```diff
@@ -0,0 +1,123 @@
+  silent: 0,
+  error: 1,
+  warn: 2,
+  info: 3,
```
Feels like overkill. I was thinking we could use console.log by default and allow for people to pass their own logger

ludamad commented:
I don't see a way for users to customize their logger with this mechanism? We should still allow someone using pino to hook it in. The point is to do it via dependency injection. We likely need to post the log messages back over the worker boundary

johnathan79717 committed:
fix(bb.js): output logger to stderr instead of stdout

Logger output should go to stderr instead of stdout to avoid interfering
with command output parsing.

This fixes an issue where BB initialization logs were being captured by
scripts that parse stdout, such as:
  aztec get-canonical-sponsored-fpc-address | awk '{print $NF}'

The logger message "[bb_sync] Using native shared memory backend: <path>"
was being captured instead of the actual command output, causing the BB
binary path to be passed as an argument to subsequent commands, leading
to "Invalid Aztec address" errors.

By using stderr for logging:
- Logger output doesn't interfere with stdout-based command pipelines
- Scripts parsing stdout work correctly
- Users still see log messages in their terminal
- Logs can be suppressed with 2>/dev/null if needed

johnathan79717 commented:
> I don't see a way for users to customize their logger with this mechanism? We should still allow someone using pino to hook it in. The point is to do it via dependency injection. We likely need to post the log messages back over the worker boundary

[Barretenberg.new](https://github.com/AztecProtocol/aztec-packages/blob/e04fcbea7056fc6caa101b38996d03aaeb498635/barretenberg/ts/src/barretenberg/index.ts#L38) already provides a `logger` option that let's you pass in your own logger. It's used by [`acir_tests/browser-test-app`](https://github.com/AztecProtocol/aztec-packages/blob/8343171352677f816e3bb54734694ccd2872b89e/barretenberg/acir_tests/browser-test-app/src/index.ts#L34) although it only passes in `console.log`. I can update the test app to demonstrate passing in pino logger works.

johnathan79717 committed:
test(bb.js): use pino logger in browser test app

Update browser-test-app to pass pino logger to bb.js instead of
console.log. This demonstrates how users can provide their own logger
implementation and ensures bb.js internal logs are routed through the
application's logging system.

Changes:
- Create bbLogger wrapper that routes to pino with source tag
- Pass bbLogger to all Barretenberg.new() calls
- Maintains structured logging throughout the application

johnathan79717 committed:
fix(bb.js): ensure browser logger outputs to console for CI benchmarks

The CI benchmark script parses plain-text console output to extract memory
usage in the format "(mem: X.XMiB)". Update bbLogger to output to both pino
(for structured logging) and console.log (for CI script parsing) to ensure
benchmark scripts can successfully extract memory metrics.

ludamad commented:
I didn't see the mechanism to capture browser logs, but if it exists then all good. And I guess it must exist... A test that can just capture some logs to an array and make sure something comes from browser is enough

johnathan79717 committed:
test(bb.js): add browser console capture verification test

Add test command that instantiates real Barretenberg instance in browser and
verifies that console logs (including memory stats) are captured by Playwright.

This test confirms that:
- bbLogger outputs to both pino and console.log
- The console.log output is captured by Playwright's console event
- CI benchmark scripts can successfully parse memory usage from logs

Test output shows real bb.js logs like:
  'Initialized BN254 CRS from memory (mem: 338.13MiB)'

Addresses review feedback to verify the actual logging mechanism works.

johnathan79717 commented:
> I didn't see the mechanism to capture browser logs, but if it exists then all good. And I guess it must exist... A test that can just capture some logs to an array and make sure something comes from browser is enough

Done in https://github.com/AztecProtocol/aztec-packages/pull/18582/commits/7d2577b16c642a1c323670b25ea6a313a24af49d

johnathan79717 commented:
cc @saleel
